internal/proxy/ui_requests.go

```diff
@@ -11,212 +11,274 @@ import (
 		if err != nil {
-			p.HandleError(w, err, requestLogEntry)
+			p.HandleError(w, err, nsLogger)
 			return
 		}
-		requestLogEntry.WithField("ns", ns).Debug("Loaded OSO token")
+		nsLogger.Info("Fetched OSO token from OSIO token")
 
-		statusCode, cookies, err := p.loginJenkins(pci, osoToken, requestLogEntry)
+		status, jenkinsCookies, err := p.loginJenkins(pci, osoToken, nsLogger)
 		if err != nil {
-			p.HandleError(w, err, requestLogEntry)
+			p.HandleError(w, err, nsLogger)
 			return
 		}
-		if statusCode == http.StatusOK {
-			for _, cookie := range cookies {
-				if cookie.Name == CookieJenkinsIdled {
-					continue
-				}
-				http.SetCookie(w, cookie)
-				if strings.HasPrefix(cookie.Name, SessionCookie) { //Find session cookie and use it's value as a key for cache
-					p.ProxyCache.SetDefault(cookie.Value, pci)
-					requestLogEntry.WithField("ns", ns).Infof("Cached Jenkins route %s in %s", pci.Route, cookie.Value)
-					requestLogEntry.WithField("ns", ns).Infof("Redirecting to %s", redirectURL.String())
-					//If all good, redirect to self to remove token from url
-					http.Redirect(w, r, redirectURL.String(), http.StatusFound)
-					return
-				}
 
-				//If we got here, the cookie was not found - report error
-				p.HandleError(w, fmt.Errorf("could not find cookie %s for %s", SessionCookie, pci.NS), requestLogEntry)
-			}
-		} else {
-			p.HandleError(w, fmt.Errorf("could not login to Jenkins in %s namespace", ns), requestLogEntry)
+		nsLogger.Infof("Jenkins Login returned: %v", cookieNames(jenkinsCookies))
+
+		if status != http.StatusOK {
+			p.HandleError(w, fmt.Errorf("could not login to Jenkins in %q namespace", ns), nsLogger)
+			return
+		}
+
+		// there could be old session cookies, so lets clear it
+		expireCookiesMatching(w, r, isSessionOrIdledCookie)
+
+		// set all cookies that we got from jenkins
+		jsessionCookie := setJenkinsCookies(w, jenkinsCookies)
+		if jsessionCookie == nil {
+			// for some reason, login didn't return a session cookie
+			p.HandleError(w, fmt.Errorf("could not find cookie %q for %q", SessionCookie, ns), nsLogger)
+			return
 		}
+
+		// Update proxy-cache to associate pci with the session cookie
+		// the cache so that, the subsequent request that would contain the
+		// the jession cookie can be used to lookup the cache
+		p.ProxyCache.SetDefault(jsessionCookie.Value, pci)
+		nsLogger.Infof("Cached Jenkins route %q in %q", pci.Route, jsessionCookie.Value)
+
+		// If all good, redirect to self to remove token from url
+		nsLogger.Infof("Redirecting to %q", redirectURL)
+		http.Redirect(w, r, redirectURL.String(), http.StatusFound)
+		return
 	}
 
-	if len(r.Cookies()) > 0 { //Check cookies and proxy cache to find user info
+	if len(r.Cookies()) > 0 {
+		// Check cookies and proxy cache to find user info
+		cookieLogger := logger.WithField("part", "cookie")
+
 		for _, cookie := range r.Cookies() {
+
 			cacheVal, ok := p.ProxyCache.Get(cookie.Value)
```

Comment:
Shouldn't we check cache only if cookie.Name == "JenkinsIdled" or cookie.Name== "JSESSIONID.xyz"? This is checking for every cookie.

Reply:
yes will fix, we can `continue` if cookie isn't isSessionOrIdled.

Do note that the look up for hash is O(1) already so isn't a huge penalty that we are paying but the check does improve the readability thou.